codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttProperties.java

```diff
@@ -0,0 +1,209 @@
+        SUBSCRIPTION_IDENTIFIER(0x0B),
+
+        // UTF-8 Encoded String properties
+        CONTENT_TYPE(0x03),
+        RESPONSE_TOPIC(0x08),
+        ASSIGNED_CLIENT_IDENTIFIER(0x12),
+        AUTHENTICATION_METHOD(0x15),
+        RESPONSE_INFORMATION(0x1A),
+        SERVER_REFERENCE(0x1C),
+        REASON_STRING(0x1F),
+        USER_PROPERTY(0x26),
+
+        // Binary Data
+        CORRELATION_DATA(0x09),
+        AUTHENTICATION_DATA(0x16);
+
+        private final int value;
+
+        MqttPropertyType(int value) {
+            this.value = value;
+        }
+
+        public int value() {
+            return value;
+        }
+
+        public static MqttPropertyType valueOf(int type) {
+            for (MqttPropertyType t : values()) {
+                if (t.value == type) {
+                    return t;
+                }
+            }
+            throw new IllegalArgumentException("unknown property type: " + type);
+        }
+    }
+
+    public static final MqttProperties NO_PROPERTIES = new MqttProperties();
+
+    public abstract static class MqttProperty<T> {
+        final T value;
+        final int propertyId;
+
+        public MqttProperty(int propertyId, T value) {
+            this.propertyId = propertyId;
+            this.value = value;
+        }
+    }
+
+    public static final class IntegerProperty extends MqttProperty<Integer> {
+
+        public IntegerProperty(int propertyId, Integer value) {
+            super(propertyId, value);
+        }
+    }
+
+    public static final class StringProperty extends MqttProperty<String> {
+
+        public StringProperty(int propertyId, String value) {
+            super(propertyId, value);
+        }
+    }
+
+    public static final class StringPair {
+        public final String key;
+        public final String value;
+
+        public StringPair(String key, String value) {
+            this.key = key;
+            this.value = value;
+        }
+
+        @Override
+        public int hashCode() {
+            return key.hashCode() + 31 * value.hashCode();
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj) {
+                return true;
+            }
+            if (obj == null || getClass() != obj.getClass()) {
+                return false;
+            }
+            StringPair that = (StringPair) obj;
+
+            return that.key.equals(this.key) && that.value.equals(this.value);
+        }
+    }
+
+    //User properties are the only properties that may be included multiple times and
+    //are the only properties where ordering is required. Therefore, they need a special handling
+    public static final class UserProperties extends MqttProperty<List<StringPair>> {
+        public UserProperties() {
+            super(MqttPropertyType.USER_PROPERTY.value, new ArrayList<StringPair>());
+        }
+
+        public UserProperties(List<StringPair> values) {
+            this();
```

Comment:
not need to call `this()`

Reply:
If I remove `this()` then it tries to call superclass default construction which doesn't exist, so compilation fails.

Reply:
ah I see...  Plese also make it clear via java docs that the given values are copied and so the List reference is not stored. You may even use `Collection<StringPair>` or `Iterable<StringPair>`